src/ggml-rpc.cpp

```diff
@@ -0,0 +1,622 @@
+    ggml::FreeBufferReply reply;
+    grpc::ClientContext context;
+    grpc::Status status = ctx->stub->FreeBuffer(&context, request, &reply);
+    GGML_ASSERT(status.ok());
+    delete ctx;
+}
+
+GGML_CALL static void * ggml_backend_rpc_buffer_get_base(ggml_backend_buffer_t buffer) {
+    GGML_PRINT_DEBUG("buffer get base\n");
+    ggml_backend_rpc_buffer_context * ctx = (ggml_backend_rpc_buffer_context *)buffer->context;
+    ggml::BufferGetBaseRequest request;
+    request.set_bufptr(ctx->remote_ptr);
+    ggml::BufferGetBaseReply reply;
+    grpc::ClientContext context;
+    grpc::Status status = ctx->stub->BufferGetBase(&context, request, &reply);
+    GGML_ASSERT(status.ok());
+    return reinterpret_cast<void *>(reply.baseptr());
+}
+
+static void serialize_tensor(const ggml_tensor * tensor, ggml::Tensor * protobuf_tensor) {
+    protobuf_tensor->set_id(reinterpret_cast<uint64_t>(tensor));
+    protobuf_tensor->set_type(tensor->type);
+    protobuf_tensor->set_backend(tensor->backend);
+    if (tensor->buffer) {
+        ggml_backend_buffer_t buffer = tensor->buffer;
+        ggml_backend_rpc_buffer_context * ctx = (ggml_backend_rpc_buffer_context *)buffer->context;
+        protobuf_tensor->set_bufptr(ctx->remote_ptr);
+    } else {
+        protobuf_tensor->set_bufptr(0);
+    }
+    for (uint32_t i = 0; i < GGML_MAX_DIMS; i++) {
+        protobuf_tensor->add_ne(tensor->ne[i]);
+    }
+    protobuf_tensor->set_op(tensor->op);
+    for (uint32_t i = 0; i < GGML_MAX_OP_PARAMS / sizeof(int32_t); i++) {
+        protobuf_tensor->add_op_params(tensor->op_params[i]);
+    }
+    protobuf_tensor->set_flags(tensor->flags);
+    for (uint32_t i = 0; i < GGML_MAX_SRC; i++) {
+        protobuf_tensor->add_src(reinterpret_cast<uint64_t>(tensor->src[i]));
+    }
+    protobuf_tensor->set_view_src(reinterpret_cast<uint64_t>(tensor->view_src));
+    protobuf_tensor->set_view_offs(tensor->view_offs);
+    protobuf_tensor->set_data(reinterpret_cast<uint64_t>(tensor->data));
+    protobuf_tensor->set_name(tensor->name);
+    protobuf_tensor->set_extra(reinterpret_cast<uint64_t>(tensor->extra));
+}
+
+static ggml_tensor * deserialize_tensor(struct ggml_context * ctx, const ggml::Tensor & protobuf_tensor) {
+    ggml_tensor * result = ggml_new_tensor_4d(ctx, (ggml_type) protobuf_tensor.type(),
+        protobuf_tensor.ne(0), protobuf_tensor.ne(1), protobuf_tensor.ne(2), protobuf_tensor.ne(3));
+    result->backend = (ggml_backend_type) protobuf_tensor.backend();
+    result->buffer = reinterpret_cast<ggml_backend_buffer_t>(protobuf_tensor.bufptr());
+    result->op = (ggml_op) protobuf_tensor.op();
+    for (uint32_t i = 0; i < GGML_MAX_OP_PARAMS / sizeof(int32_t); i++) {
+        result->op_params[i] = protobuf_tensor.op_params(i);
+    }
+    result->flags = protobuf_tensor.flags();
+    result->data = reinterpret_cast<void *>(protobuf_tensor.data());
+    strncpy(result->name, protobuf_tensor.name().c_str(), GGML_MAX_NAME);
```

Comment:
In practice this will work as long as the server and the client have the same `GGML_MAX_NAME`, but it is not safe to use `strncpy` in this way because it does not guarantee that the string will be NUL-terminated. `snprintf` should be safe.